onnxruntime/core/graph/contrib_ops/internal_nhwc_onnx_opset.cc

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "internal_nhwc_onnx_opset.h"
+
+#include "onnx/defs/operator_sets.h"
+
+#include "core/graph/constants.h"
+#include "core/graph/contrib_ops/nhwc_inference_context.h"
+
+namespace onnxruntime {
+namespace internal_nhwc_onnx {
+using contrib::NhwcInferenceContext;
+
+template <typename F>
+void RegisterNHWCSchema(F&& f, ::ONNX_NAMESPACE::OpSchema&& schema) {
+  // Need to copy the inferencing function from the temporary OpSchema object
+  auto onnx_inferencing_func = schema.GetTypeAndShapeInferenceFunction();
+  f(std::move(::ONNX_NAMESPACE::OpSchema(schema)
+                  .TypeAndShapeInferenceFunction([onnx_inferencing_func](ONNX_NAMESPACE::InferenceContext& ctx) {
+                    // use the NHWC inferencing context to convert input 0 and output 0 to NCHW
+                    // so the ONNX shape inferencing can be used. Once that completes, the call to PropagateOutputShape
+                    // will convert the inferred shape from NCHW to NHWC
+                    NhwcInferenceContext nhwc_ctx(ctx);
+                    onnx_inferencing_func(nhwc_ctx);
+                    nhwc_ctx.PropagateOutputShape();
+                  })
+                  .SetDomain(onnxruntime::kMSInternalNHWCDomain)));
+}
+
+template <typename F>
+void RegisterNHWCConvWithActivation(F&& f, ::ONNX_NAMESPACE::OpSchema&& schema) {
+  auto onnx_inferencing_func = schema.GetTypeAndShapeInferenceFunction();
+  f(std::move(::ONNX_NAMESPACE::OpSchema(schema)
+                  .Attr("activation", "", ONNX_NAMESPACE::AttributeProto::STRING, ONNX_NAMESPACE::OPTIONAL_VALUE)
+                  .Attr("activation_params", "", ONNX_NAMESPACE::AttributeProto::FLOATS, ONNX_NAMESPACE::OPTIONAL_VALUE)
+                  .TypeAndShapeInferenceFunction([onnx_inferencing_func](ONNX_NAMESPACE::InferenceContext& ctx) {
+                    NhwcInferenceContext nhwc_ctx(ctx);
+                    onnx_inferencing_func(nhwc_ctx);
+                    nhwc_ctx.PropagateOutputShape();
+                  })
+                  .SetDomain(onnxruntime::kMSInternalNHWCDomain)));
+}
+
+#define REGISTER_NHWC_SCHEMA(Op, SinceVersion) \
```

Comment:
>REGISTER_NHWC_SCHEMA [](http://example.com/codeflow?start=8&length=20)

nit: consider also passing in `fn` as a macro parameter so there's less contextual dependency.
also, do we need to introduce a separate scope? if so, `do { ... } while(0)` has the additional benefit of requiring a semi-colon at the end

Reply:
Done. Not sure we need a new scope so removing that.